content/ADR-253-light-sources.md

```diff
@@ -0,0 +1,117 @@
+---
+layout: adr
+adr: 253
+title: Light Sources
+date: 2024-11-13
+status: Draft
+type: RFC
+spdx-license: CC0-1.0
+authors:
+  - nearnshaw
+---
+
+# Abstract
+
+This document explores a new feature that has been missing from Decentraland scenes: letting creators control light. We will allow creators to define a limited number of entities in their scenes that will behave as sources of light, and offer a number of parameters to control the kind of light-source, color, luminosity, etc.
+
+We also explore how we can make this work in an open world, and its implications.
+We also discuss limitations that should be imposed, to prevent a degradation of the experience if this feature is misused.
+
+# Context
+
+Up until now, all light in Decentraland came from a single rigid source (the sun or moon). Enabling creators to create their own light sources enables a lot of creative possibilities. We see the control of light as important for the following:
+
+- Creating ambience and enhance the architecture of a scene
+- Flashy effects for music festivals, fashion week, and other events of high visual impact
+- Light as a visual cue: guide the player towards what’s important in a scene by shining a light on it. Signal a change in what’s going on in the scene by switching the lights to a different intensity or color.
+- Darkness as a mechanic for spooky games
+- Flashlights as a mechanic: the act of having to point a light to reveal what’s there can be a whole game mechanic.
+
+To be able to play with darkness, we'll also need to provide controls to be able to disable the global default light sources and play with that.
+
+Note: Some GLTF models come with lights packaged as nodes inside the structure of the model. We should ignore those completely. The only lights we will be rendering are the ones defined via code using the LightSource component.
+
+## Component description
+
+We should create a `LightSource` component that, when added to an entity, tells the engine to shine a light from that entity’s Transform position.
+
+A Type field will let you chose between _Spot_ and _Point_ light. We believe these two types of lights are enough for now, the component could be expanded in the future if we want to include other types.
```

Comment:
Regarding the component structure, I'd like to discuss an alternative approach based on composition rather than a single component with type discrimination:

Instead of:
```typescript
interface LightSource {
  type: 'point' | 'spot' | 'directional';
  intensity: number;
  // ... common properties ...
  // spot-specific properties when type === 'spot'
  angle?: number;
  innerAngle?: number;
}
```

Consider a composable approach:
```typescript
interface Light {
  illuminance: number;
  color: Color3;
  shadows: boolean;
}

interface Spotlight {
  angle: number;
  innerAngle?: number;
}

interface GlobalLight {
  direction?: Vector3;
  ambientColor?: Color3;
  ambientBrightness?: number;
}
```

Benefits:
1. **Cleaner Separation**: Each component handles one specific aspect of lighting
2. **Type Safety**: No need for runtime type checking or optional properties
3. **Extensibility**: New light types can be added without modifying existing components
4. **Flexibility**: SDKs can provide their own abstractions while the protocol remains clean
5. **Clarity**: Entity behavior is determined by component composition:
   - Point Light = Entity + Light
   - Spot Light = Entity + Light + Spotlight
   - Directional Light = Root Entity + Light + GlobalLight

Reply:
We debated this point, and reached the conclusion that we still much prefer to use a single component, for the following reasons:

1. It's consistent with how many other components in our SDK already work (eg Material, Tween, MeshRenderer, MeshCollider)
2. It's friendlier to have to remember a single component name, and have the field suggestions do the rest for you. If you need to add two separate components, you don't have that guidance, you need to remember the two component names by heart. Users can also get potentially lost if they try to only add the SpotLight component and see no effect.
3. We can add helpers, like we already do on other components to make this friendlier. eg:
`LightSource.spotLight(myEntity, { brightness: 1000, angle: 10})`